### ISSUE TITLE
[WritableStream] Shouldn't the state after writer.abort() be closed?

### ISSUE THREAD

**Issue by tyoshino** [whatwg/streams#464]
For ReadableStream, `reader.cancel(reason)` passes `reason` to the underlying source via `source.cancel()`, but we decided not to have it set to storedError and closed but have the stream just say it's cleanly closed.

On the other hand, `writer.abort()` is specified to set storedError and closed to return the passed reason.

IIRC, this was required long time ago for aborting pipeTo and having the source cancelled as if the destination is errored with the reason passed to the abort() call. It's now impossible given the writer/reader structure.

I've been feeling a bit uncomfortable with the error mirroring (both to sink and WritableStream I/F). If it's true that we no longer have any justification for it, I'd like to revisit it and consider removal.

**Comment by domenic** [comment]
> IIRC, this was required long time ago for aborting pipeTo and having the source cancelled as if the destination is errored with the reason passed to the abort() call. It's now impossible given the writer/reader structure.

Can you explain why it is impossible now? I don't see why things would have changed. Error propagation through the pipe chain is super important; we can't lose that. The lack of it in Node streams has been very painful.

---

So I have always envisioned readable stream canceling and writable stream aborting as conceptually quite different, and not mirrors. I touched on this a little in https://github.com/whatwg/streams/blob/master/Requirements.md#you-must-be-able-to-communicate-cancel-signals-up-a-pipe-chain:

> This signal is similar to, but different than, the "close" signal described above. It implies that the data written so far is not meaningful or valid; that any queued writes should be discarded; and that a cleanup operation should be performed. For example, when a file stream represents a newly-created file, a close signal waits for all queued writes to complete, then closes the file descriptor; a new, complete file now exists on the disk. But a abort signal would throw away any queued writes and delete the file.

I think we can both agree that it's useful to have a way (currently called "abort") that throws away all written data.

I guess the question is, after throwing away all data, should your state be "errored" (with the passed error)? Or should we reserve "errored" only for when writing fails, and say that this is more of a "premature close"? (We could even have separate APIs, `.discardQueue()` and `.close()`.) I would be OK reconsidering this, but am not convinced either way yet.

**Comment by tyoshino** [comment]
> Can you explain why it is impossible now? I don't see why things would have changed. Error propagation through the pipe chain is super important; we can't lose that. The lack of it in Node streams has been very painful.

Oh, sorry for lack of context. I was talking about aborting pipeTo() by abort()-ing the destination. It's impossible as we're adding the locking mechanism to the writable stream suite, and so the destination writable stream is locked by pipeTo(). No one other than the ongoing pipeTo() may abort the writable stream.

At some point, IIRC, we discussed some redesign of pipeTo() interface. After introducing locking mechanism to both, we need to extend pipeTo() to give us some controller (I'm not catching up with progress on cancellable promise. It could be to some extent covered by it).

The controller may have a method for abort()-ing the pipeTo(). It would take an object to pass to source.cancel() and dest.abort(). We could design it to take the errors for them separately or only one for both.
- pipeController.abort(sourceReason, destReason);
- pipeController.abort(reason);

**Comment by tyoshino** [comment]
Hmm, yeah. I remembered that we agreed on that abort() should play the role of having a writable stream pretend to be errored by itself for the externally-provided reason. But I thought that it was backed by the demand to implement the way to abort pipeTo() operation and propagating error to both direction. It's gone.

**Comment by domenic** [comment]
> At some point, IIRC, we discussed some redesign of pipeTo() interface. After introducing locking mechanism to both, we need to extend pipeTo() to give us some controller (I'm not catching up with progress on cancellable promise. It could be to some extent covered by it).

Right. FYI this will probably be covered by passing a cancel token as an option to pipeTo. So

```js
const { token, cancel } = CancelToken.source();
rs.pipeTo(ws, { cancelToken: token });

cancelButton.onclick = cancel;
```

This would cancel `rs`, clearly, but I am not sure whether it should abort or close `ws`. Hmm. I guess that could even be an option that you pass to `cancel()`.

> Hmm, yeah. I remembered that we agreed on that abort() should play the role of having a writable stream pretend to be errored by itself for the externally-provided reason. But I thought that it was backed by the demand to implement the way to abort pipeTo() operation and propagating error to both direction. It's gone.

I don't really remember that. To me the important thing is that if you do something like

```js
rs.pipeThrough(transform).pipeTo(ws);
```

and then writing to `ws` errors (I/O error etc.), this propagates backward. At the very least that means `rs` must be canceled. But I think it should also mean that `transform.writable` becomes errored and `transform.readable` becomes closed?

The way I originally envisioned this working is that pipeTo would see `ws` become errored, and use that to cancel `transform.readable`, and then the transform stream machinery would see the premature cancel and react to it to make `transform.writable` errored, which would mean that the piping of `rs` to `transform.writable` sees its destination error and cancels `rs`. This seems nice to me.

But maybe it is not the optimal design and I should be more creative thinking of ways for this to work. I do not think it is inherently important that `transform.writable` become errored and that `transform.readable` become closed. That was just a nice way to accomplish the important goal that `rs` become canceled.

**Comment by tyoshino** [comment]
> This would cancel rs, clearly, but I am not sure whether it should abort or close ws. Hmm. I guess that could even be an option that you pass to cancel().

Agreed. Similar to normal completion case for which we already have prevent.* options.

> The way I originally envisioned ...

Yeah, that's not different from one in my mind. I posted [this](https://github.com/whatwg/streams/issues/464#issuecomment-223197040) just to explain how (I remember) I was convinced by you with why `ws.abort()` should make `ws` errored. My proposal doesn't break error propagation by `pipeTo()` itself. Left to right propagation is done by `writer.abort()` call on transform.writable. The transform stream machinery would forward the signal to `pipeTo()` machinery by erroring `transform.readable`. It'll be propagated to result in `ws.abort()` call. My proposal doesn't affect this.

> I do not think it is inherently important that transform.writable become errored and that transform.readable become closed.

Agreed. In general, to what state (state visible on WritableStream I/F. Not about the sink and internal state) `ws` transitions to on `writer.abort()` and what state (state visible on ReadableStream I/F. Not about the source and internal state) `rs` transitions to on `reader.cancel()` are not so important. Consumer and producer facing the stream interface should know what they did. So, my proposal doesn't solve any real issue but kinda based on preference and consistency.

**Comment by domenic** [comment]
OK. So there are four possible things you might want to do as a producer to terminate your relationship with a writable stream:

- (A) queue up a close, so that after all pending writes finish, the stream goes to the closed state
- (B) queue up an error, so that after all pending writes finish, the stream goes to the errored state
- (C) immediately close. Throw away any queued writes, and then move the stream to the closed state
- (D) immediately error. Throw away any queued writes, and then move the stream to the errored state

Let's also write down what "closed" vs. "errored" means. The difference is kind of subtle.

- If a writable stream is closed:
  - `write()` will reject with a TypeError
  - `abort()` will be a no-op and return a promise fulfilled with undefined
  - `closed` will fulfill
  - `waitForDesiredSize()` will stay pending forever?
  - desiredSize will be 0?
  - `getWriter()` will return a writer with the above behaviors?
- If a writable stream is errored:
  - `write()` will reject with the stored error
  - `abort()` will reject wit the stored error
  - `closed` will reject with the stored error
  - `waitForDesiredSize()` will reject with the stored error
  - `desiredSize` will be null??
  - `getWriter()` will return a writer with the above behaviors?

The overall theme here is that any interactions with an errored writable stream will surface that error. So if you were expecting to do anything further with it, you will instead get an error, and hopefully surface it to your user/logs/etc. Whereas with a closed writable stream ... mostly nothing will happen? (`write()` is kind of an exception).

I am not sure the distinction here is a great distinction. I guess it is strongest for the `.closed` promise, which is useful, because of code like this:

```js
writer.write(a);
writer.write(b);
writer.write(c);
writer.close();

writer.closed.then(
  () => console.log("everything succeeded"),
  err => console.error("something went wrong in one of the writes or in closing", e)
);
```

---

Anyway, with this as background: My point of view is that (A) is very common, and that among the remaining possibilities, (D) is also common.

I think you are arguing that (C) is more common (or maybe just more consistent/preferred) than (D). Is that correct?

To me (D) seems more likely, because if you are throwing away queued data, something has likely gone wrong, and other parts of your program (mostly those that use `writer.closed`) should probably go down the error-handling path.

What do you think, given this framing of the problem?

**Comment by tyoshino** [comment]
(a) I agree that it's more natural that an aborted writable stream says "hey, I'm broken" than "I'm closed". It's good to have them behave nicely so that developers who made mistake get the most useful signal from them.

But we can have similar argument against `readableStream.cancel()`. Readable stream's "closed" state means end-of-data for normal cases. A cancelled stream is not telling us that it reached the end-of-data.

(b) If we have any logic that tries to do something further after .abort(), .cancel(), that's just wrong. If there's a fulfillment callback set to readableStream.closed, it should be neutered when we call readableStream.cancel(). Similarly, any code that is waiting for writableStream should be neutered when we're calling writableStream.abort().

> if you are throwing away queued data, something has likely gone wrong, and other parts of your program (mostly those that use writer.closed) should probably go down the error-handling path.

This makes me feel that an aborted writable stream looks more inappropriate target to touch than a cancelled readable stream, yeah ... but ... very very subtle. Both are basically wrong.

----

The current setup has 2 points look strange to me:
- (1) inconsistency between abort() and cancel(). cancel() is backed by (b) but abort() is backed by (a)
- (2) the reason passed to .abort() is set to storedError and reflected back to the producer.

(i) Ideally I'd like to change cancel() to enter "cancelled" state and make methods throw/reject with TypeError("Cancelled") and make abort() to enter "aborted" state and make methods throw/reject with TypeError("Aborted"). This solves both (1) and (2) and satisfies (a). As we're likely not to expose the state, there wouldn't be any additional complexity (additional state "aborted" and "cancelled"). but we cannot revert .cancel()'s behavior now. It's already shipped. Too late.

(ii) Changing abort() to transition to closed solves (1) and (2) but doesn't satisfy (a).

(iii) Changing abort(reason) not to set reason to storedError but set it to TypeError("Aborted") solves (2) and satisfies (a).

I've been suggesting (ii) but could also live with (iii) honoring importance of (a).

**Comment by isonmad** [comment]
> but we cannot revert .cancel()'s behavior now. It's already shipped. Too late.

You mean the implementation shipping [since Chrome 43](https://www.chromestatus.com/feature/4531143755956224), right? How many others have implemented it yet, aside from [IE10 shipping an incompatible API](https://www.chromestatus.com/feature/6605041225957376); is that not early enough for a breaking change still?

**Comment by tyoshino** [comment]
> You mean the implementation shipping since Chrome 43, right?

Right. Firefox hasn't yet shipped it. We may change the behavior now as it's subtle though it's been there for ~1 y.

**Comment by domenic** [comment]
> But we can have similar argument against readableStream.cancel(). Readable stream's "closed" state means end-of-data for normal cases. A cancelled stream is not telling us that it reached the end-of-data.

I think the difference is in consumer vs. producer. If a consumer is no longer interested in the stream, then it is not an error: nothing is broken. Everything is going as expected. A consumer is no longer interested in data in two cases: one, when there is no more data. And two, when they explicitly signal disinterest. Thus, cancel() sends you to the closed state.

Whereas if a producer can no longer give any data (and wants to throw away any queued data), then something is usually broken.

> If we have any logic that tries to do something further after .abort(), .cancel(), that's just wrong. If there's a fulfillment callback set to readableStream.closed, it should be neutered when we call readableStream.cancel().

I tend to agree with this. This is what motivates my ["third state" thinking for cancelable promises](https://github.com/domenic/cancelable-promise/blob/master/Third%20State.md), which would cause both fulfillment and rejection callbacks to be skipped. However, the timing is pretty poor here. We could potentially consider changing the semantics of `.cancel()` to cancel the `.closed` promise (instead of fulfilling it), but it would be backward-incompatible. I didn't want to introduce a path dependency on cancelable promises. Especially since they are controversial and the third state idea might not be able to make it through committee :(.

> Similarly, any code that is waiting for writableStream should be neutered when we're calling writableStream.abort().

I think this is a bit different. To me it makes sense that any code that is waiting for the writable stream that gets aborted should be told about abort as an error. You abort() because something broke. It is often good to learn about what broke and surface that error.

I agree it's a subtle distinction.

> The current setup has 2 points look strange to me:

To me these are not very big worries. (1) inconsistency doesn't worry me because we specifically named them differently. And (2) doesn't seem like a bad thing; I don't understand why it's a problem.

> (i) Ideally I'd like to change cancel() to enter "cancelled" state and make methods throw/reject with TypeError("Cancelled") and make abort() to enter "aborted" state and make methods throw/reject with TypeError("Aborted").

I think this is a bad user experience for people using readable streams. Imagine writing code like `rs.closed.catch(tellUserTheDownloadFailed)`. Then the user navigates to another part of your webapp, which means the download doesn't matter any more, so you do `rs.cancel()`. Then the user gets a popup saying the download failed, which is a bad user experience. It's similar to my arguments against cancelation as an error [in the cancelable promises repo](https://github.com/domenic/cancelable-promise/blob/master/Third%20State.md). (Maybe cancelation as success isn't super-great either, but I think for the particular case of streams it works pretty OK.)

> (ii) Changing abort() to transition to closed solves (1) and (2) but doesn't satisfy (a).

I don't object _too_ strongly to this. In this case we would rename `abort()` to `cancel()` and treat it as (C) from my earlier post. ("immediately close. Throw away any queued writes, and then move the stream to the closed state").

It comes down to the subtle issue of whether you believe that when  producer can no longer give any data (and wants to throw away any queued data), then something is usually broken. If you think that producers actually want to do that for other reasons than things being broken, then maybe (ii) is OK. But I think usually in this case something will be broken.

> (iii) Changing abort(reason) not to set reason to storedError but set it to TypeError("Aborted") solves (2) and satisfies (a).

I would be OK with this, but I don't understand why (2) is a problem we want to solve. It seems like a reasonable thing to say "the reason the stream is broken is because `reason`" and that further interactions with the stream will tell you that. In other words, it sounds like you would prefer "the reason the stream is broken is because it was aborted" whereas I am thinking it's useful to say "the reason the stream is broken is because `reason` (which is why we aborted it)".

**Comment by tyoshino** [comment]
> I think the difference is in consumer vs. producer. ...

I agree with you on this point. Unfinished WS is likely to be broken while unfinished RS is not broken.

> A consumer is no longer interested in data in two cases: one, ...

They're no longer interested in data after seeing closure, I agree. But transition itself often has some meaning, i.e. processed data FULLY. That's the main point I used for the counter argument.

> I think this is a bad user experience for people using readable streams. Imagine writing code like rs.closed.catch(tellUserTheDownloadFailed). Then the user navigates to another part of your webapp, which means the download doesn't matter any more, so you do rs.cancel(). Then the user gets a popup saying the download failed, which is a bad user experience. It's similar to my arguments against cancelation as an error in the cancelable promises repo. (Maybe cancelation as success isn't super-great either, but I think for the particular case of streams it works pretty OK.)

Hmm, so, there's philosophy behind design of closed that rejection of closed should not depend on what actions we made on the RS? I agree that less path-dependency is good. But for fulfillment, we now have two paths to reach it. In rs.closed.then(tellUserTheDownloadSucceeded), tellUserTheDownloadSucceeded should know whether cancel has happened or not to do the right thing (show popup or do nothing).

So, I'm ... convinced with that rejecting closed for cancel() is bad as well as fulfilling closed. You're also saying "Maybe cancelation as success isn't super-great...". Yeah.

> I think this is a bit different. To me it makes sense that any code that is waiting for the writable stream that gets aborted should be told about abort as an error. You abort() because something broke. It is often good to learn about what broke and surface that error.

I totally understand your logic. But the only disagreement is that having two different paths reach to fulfillment of closed on RS is fine while having two different paths "we aborted" and "WS errored" reach to rejection of closed on WS is bad.

>  And (2) doesn't seem like a bad thing; I don't understand why it's a problem.

To surface an error, we need to know both who's the culprit and the type of error. We can of course include (a hint of) who's the culprit in the exception by e.g. having a file reading RS error using a descriptive error type, say, FileReadableStreamFailedByXXXException while having a file writing WS error FileWritableStreamFailedByXXXException. With this, a rejection callback set to ws.closed can show the right message (right culprit info) to user (or some other code) just by investigating the exception. It's not needed to know which path it has taken (called abort() or not).

But this is redundant. For example, XXX can be common error such as "OutOfMemory". We can just use a single OutOfMemoryException across streams if we don't rely on the requirement above.

I'm leaning toward forcing users to have their code interact with a WS while understanding what it did (cancel()-ed or not, etc.) to know in which path it is on to correctly interpret signal from WS (closed, etc.), than relying on that streams include (hint for knowing) culprit information in the exceptions.

> In other words, it sounds like you would prefer "the reason the stream is broken is because it was aborted" whereas I am thinking it's useful to say "the reason the stream is broken is because reason (which is why we aborted it)".

Sorry, I was unclear about what I really don't like. Let me clarify. I'm uncomfortable with that an error passed to .abort() is set to storedError as-is on closed. See the OutOfMemory example above. I'm fine with including it on closed. "as-is" is the problem. For example, I'm totally fine if storedError is set to WritableStreamAbortedError and the instance has a reason field which includes the error passed to .abort().

**Comment by tyoshino** [comment]
By "redundant" in the last comment, I meant that it's already known to the code who's the culprit.

I got this feeling when updating pipeTo(). Because pipeTo() knows nothing about source and dest's error semantics, it cannot get anything from the rejection reason of dest.closed for deciding what to do in the rejection callback.

**Comment by tyoshino** [comment]
Now I'm kinda convinced with that (i) and (ii) are not the right thing to pursue given your comments and that this is a bikeshed.

Wanna know your impression on https://github.com/whatwg/streams/issues/464#issuecomment-228675316 regarding (iii).

**Comment by tyoshino** [comment]
I have similar feeling for the design of Promise.all()'s rejection. Promise.all() rejects with the reason of the first rejection. There's no index attached to it. So, if we want to show which operation has failed, we need to retrieve that info by using something else than the rejection reason of Promise.all().

I guess for Promise.all() you've dropped that info after comparing the benefit of simplicity and the info, and decided to ask users who really want to show such info to do some more work by themselves. Was there any discussion from which I can get some rationale that helps me get convinced with this? I searched for "Promise.all index" at https://github.com/domenic/promises-unwrapping/issues but couldn't find such discussion.

**Comment by domenic** [comment]
I think I would be OK with (iii). Changing it to a TypeError saying that the stream is aborted seems good. In the future we can consider adding a `.reason` property if people want it, although thinking about it, I kind of doubt they will.

I don't recall any discussions about that in Promise.all, yeah. It's more just, signaling what caused the process to fail. I think in most use cases of exceptions you need to put enough information in the exception (or implicitly in its stack trace, which is usually saved at creation time) to allow people to diagnose the problem.

I think we agree that cancelation as success was not super-great for readable streams. Maaaaybe if cancelable promises get through, and we add use counters showing people rarely reach the fulfillment handler for closed on a canceled stream, we could change it. But likely it will just be a small design wart.

**Comment by tyoshino** [comment]
Fixed by PR #488. abort() now follows (iii).
